x/src/lint/engine/runner.rs

```diff
@@ -0,0 +1,183 @@
+        self.fail_fast = fail_fast;
+        self
+    }
+
+    pub fn build(&self) -> LintEngine<'cfg> {
+        LintEngine::new(self.clone())
+    }
+}
+
+/// Executor for linters.
+#[derive(Debug)]
+pub struct LintEngine<'cfg> {
+    config: LintEngineConfig<'cfg>,
+
+    // Caches to allow results to live as long as the engine itself.
+    files_stdout: OnceCell<Vec<u8>>,
+}
+
+impl<'cfg> LintEngine<'cfg> {
+    pub fn new(config: LintEngineConfig<'cfg>) -> Self {
+        Self {
+            config,
+            files_stdout: OnceCell::new(),
+        }
+    }
+
+    pub fn run(&self) -> Result<LintResults> {
+        let mut skipped = vec![];
+        let mut messages = vec![];
+
+        // TODO: add support for project and file lints.
+
+        let project_ctx = ProjectContext::new(self.config.project_root);
+
+        if let Some(content_linters) = self.config.content_linters {
+            let file_list = self.get_file_list()?;
+
+            // TODO: This should probably be a worker queue with a thread pool or something.
+
+            let file_ctxs: Vec<_> = file_list
+                .iter()
+                .map(|path| FileContext::new(project_ctx, path))
+                .collect();
+
+            'outer: for file_ctx in file_ctxs {
+                let linters_to_run = content_linters
+                    .iter()
+                    .copied()
+                    .filter_map(|linter| match linter.pre_run(&file_ctx) {
+                        Ok(RunStatus::Executed) => Some(Ok(linter)),
+                        Ok(RunStatus::Skipped(reason)) => {
+                            let source = file_ctx.source(linter.name());
+                            skipped.push((source, reason));
+                            None
+                        }
+                        Err(err) => Some(Err(err)),
+                    })
+                    .collect::<Result<Vec<_>>>()?;
+
+                if linters_to_run.is_empty() {
+                    // No linters to run for this file -- no point loading it.
+                    continue;
+                }
+
+                // Load up the content for this file.
+                let content_ctx = file_ctx.load()?;
+
+                for linter in linters_to_run {
+                    let source = content_ctx.source(linter.name());
+                    let mut formatter = LintFormatter::new(source, &mut messages);
+
+                    match linter.run(&content_ctx, &mut formatter)? {
+                        RunStatus::Executed => {
+                            // Yay! Lint ran successfully.
+                        }
+                        RunStatus::Skipped(reason) => {
+                            skipped.push((source, reason));
+                        }
+                    }
+
+                    if self.config.fail_fast && !messages.is_empty() {
+                        // At least one issue was found.
+                        break 'outer;
```

Comment:
So if i'm correctly reading this, in order to signal that an error occurred the lint has to write into LintFormatter? I haven't thought about this much but I'm assuming this is better than signaling that via the RunStatus enum?

Reply:
Yes -- I personally really like this pattern because it cleanly separates out "this lint ran successfully but produced some errors and/or warnings" from "some sort of exceptional situation happened while handling this". Writing to a mutable vector also means that we'll avoid lots of little allocations for errors/warnings.

Reply:
Super fan of this pattern.    This is generally called a "listener", and can be made fairly generic.  I generally try to make listeners  wrap generic func signatures that can be applied around func.   No idea if that's doable in rust.   Simply returning an unused error could be captured then.

Reply:
"unused"

Reply:
I guess the trait is basically a generic func signature?